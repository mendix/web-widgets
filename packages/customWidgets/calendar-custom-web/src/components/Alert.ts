import { createElement, FunctionComponent, PropsWithChildren } from "react";
import classNames from "classnames";

export interface AlertProps {
    bootstrapStyle?: "default" | "primary" | "success" | "info" | "warning" | "danger";
    className?: string;
}

export const Alert: FunctionComponent<PropsWithChildren<AlertProps>> = ({ bootstrapStyle, className, children }) =>
    children
        ? createElement("div", { className: classNames(`alert alert-${bootstrapStyle}`, className) }, children)
        : null;

Alert.displayName = "Alert";
Alert.defaultProps = { bootstrapStyle: "danger" };
