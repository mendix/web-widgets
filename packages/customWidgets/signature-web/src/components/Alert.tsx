import { FC, PropsWithChildren } from "react";
import classNames from "classnames";

export interface AlertProps extends PropsWithChildren {
    bootstrapStyle?: "default" | "primary" | "success" | "info" | "warning" | "danger";
    className?: string;
}

export const Alert: FC<AlertProps> = ({ bootstrapStyle = "danger", className, children }) =>
    children ? <div className={classNames(`alert alert-${bootstrapStyle}`, className)}>{children}</div> : null;

Alert.displayName = "Alert";
