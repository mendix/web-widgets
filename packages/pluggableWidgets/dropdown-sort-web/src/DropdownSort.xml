<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.dropdownsort.DropdownSort" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Drop-down sort</name>
    <description />
    <studioProCategory>Data controls</studioProCategory>
    <studioCategory>Data Controls</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/modules/gallery#4-1-drop-down-sort</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="linkedDs" type="datasource" isLinked="true" isList="true">
                    <caption>Datasource to sort</caption>
                    <description />
                </property>
                <property key="attributes" type="object" isList="true" required="false">
                    <caption>Attributes</caption>
                    <description>Select the attributes that the end-user may use for sorting</description>
                    <properties>
                        <propertyGroup caption="General">
                            <property key="attribute" type="attribute" dataSource="../linkedDs" isMetaData="true" required="true">
                                <caption>Attribute</caption>
                                <description />
                                <attributeTypes>
                                    <attributeType name="AutoNumber" />
                                    <attributeType name="Decimal" />
                                    <attributeType name="Integer" />
                                    <attributeType name="Long" />
                                    <attributeType name="String" />
                                    <attributeType name="DateTime" />
                                    <attributeType name="Boolean" />
                                    <attributeType name="Enum" />
                                </attributeTypes>
                            </property>
                            <property key="caption" type="textTemplate" required="true">
                                <caption>Caption</caption>
                                <description />
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="emptyOptionCaption" type="textTemplate" required="false">
                    <caption>Empty option caption</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Accessibility">
            <propertyGroup caption="Accessibility">
                <property key="screenReaderButtonCaption" type="textTemplate" required="false">
                    <caption>Sort order button caption</caption>
                    <description>Assistive technology will read this upon reaching the sort order button.</description>
                </property>
                <property key="screenReaderInputCaption" type="textTemplate" required="false">
                    <caption>Input caption</caption>
                    <description>Assistive technology will read this upon reaching the input element.</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
