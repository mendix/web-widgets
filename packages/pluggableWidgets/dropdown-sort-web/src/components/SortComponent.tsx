import { createElement, CSSProperties, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { useOnClickOutside } from "@mendix/widget-plugin-hooks/useOnClickOutside";
import { usePositionObserver } from "@mendix/widget-plugin-hooks/usePositionObserver";
import { SortDirection } from "@mendix/widget-plugin-sorting";
import classNames from "classnames";
import { createPortal } from "react-dom";

export interface SortOption {
    caption: string;
    value: string;
}

interface SortComponentProps {
    className?: string;
    defaultDirection?: SortDirection;
    defaultOption?: SortOption;
    emptyOptionCaption?: string;
    id?: string;
    options: SortOption[];
    tabIndex?: number;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    styles?: CSSProperties;
    updateSort?: (value: SortOption, direction: SortDirection) => void;
}

export function SortComponent(props: SortComponentProps): ReactElement {
    const [valueInput, setValueInput] = useState(props.defaultOption?.caption ?? props.emptyOptionCaption ?? "");
    const [selectedSort, setSelectedSort] = useState<SortOption>(
        props.defaultOption ?? {
            caption: props.emptyOptionCaption ?? "",
            value: ""
        }
    );
    const [show, setShow] = useState(false);
    const [dropdownWidth, setDropdownWidth] = useState(0);
    const [direction, setDirection] = useState(props.defaultDirection ?? "asc");

    const componentRef = useRef<HTMLDivElement>(null);
    const optionsRef = useRef<HTMLUListElement>(null);

    const position = usePositionObserver(componentRef.current, show);

    const onClick = useCallback((option: SortOption) => {
        setValueInput(option.caption);
        setSelectedSort(option);
        setShow(false);
    }, []);

    useOnClickOutside([componentRef, optionsRef], () => setShow(false));

    useEffect(() => {
        if (selectedSort) {
            props.updateSort?.(selectedSort, direction);
        }
    }, [direction, selectedSort]);

    const showPlaceholder = !selectedSort || valueInput === props.emptyOptionCaption;

    const optionsComponent = createPortal(
        <ul
            ref={optionsRef}
            id={`${props.id}-dropdown-list`}
            className="dropdown-list"
            role="menu"
            data-focusindex={0}
            style={{ position: "fixed", width: dropdownWidth, top: position?.bottom, left: position?.left }}
        >
            {props.options.map((option, index) => (
                <li
                    className={classNames({
                        "filter-selected": selectedSort?.value === option.value
                    })}
                    key={index}
                    onClick={e => {
                        e.preventDefault();
                        e.stopPropagation();
                        onClick(option);
                    }}
                    onKeyDown={e => {
                        if (e.key === "Enter" || e.key === " ") {
                            e.preventDefault();
                            e.stopPropagation();
                            onClick(option);
                        } else if (e.key === "Tab" && index + 1 === props.options.length) {
                            e.preventDefault();
                            setShow(false);
                            componentRef.current?.querySelector("button")?.focus();
                        } else if ((e.key === "Tab" && e.shiftKey && index === 0) || e.key === "Escape") {
                            e.preventDefault();
                            setShow(false);
                            componentRef.current?.querySelector("input")?.focus();
                        }
                    }}
                    role="menuitem"
                    tabIndex={0}
                >
                    <div className="filter-label">{option.caption}</div>
                </li>
            ))}
        </ul>,
        document.body
    );

    const containerClick = useCallback(() => {
        setShow(show => !show);
        setTimeout(() => {
            (optionsRef.current?.querySelector("li.filter-selected") as HTMLElement)?.focus();
        }, 10);
    }, []);

    return (
        <div
            className={classNames("dropdown-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            ref={componentRef}
            style={props.styles}
        >
            <div className="dropdown-triggerer-wrapper">
                <input
                    value={showPlaceholder ? "" : valueInput}
                    placeholder={showPlaceholder ? props.emptyOptionCaption : undefined}
                    className="form-control dropdown-triggerer"
                    onClick={containerClick}
                    onKeyDown={e => {
                        if (e.key === "Enter" || e.key === " ") {
                            e.preventDefault();
                            e.stopPropagation();
                            containerClick();
                        }
                    }}
                    aria-haspopup
                    ref={inputRef => {
                        if (inputRef && inputRef.clientWidth) {
                            setDropdownWidth(inputRef.clientWidth);
                        }
                    }}
                    aria-expanded={show}
                    aria-controls={`${props.id}-dropdown-list`}
                    aria-label={props.screenReaderInputCaption}
                />
                <button
                    aria-label={props.screenReaderButtonCaption}
                    className={classNames("btn btn-default btn-sort", {
                        "icon-asc": direction === "asc",
                        "icon-desc": direction === "desc"
                    })}
                    onClick={() => setDirection(prev => (prev === "asc" ? "desc" : "asc"))}
                />
            </div>
            {show && optionsComponent}
        </div>
    );
}
