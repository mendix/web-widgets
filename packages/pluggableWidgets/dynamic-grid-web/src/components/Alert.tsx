import { ReactElement, createElement } from "react";
import classNames from "classnames";

export interface AlertProps {
    message?: string;
    className?: string;
    bootstrapStyle: "default" | "primary" | "success" | "info" | "inverse" | "warning" | "danger";
}

export function Alert({ message, className, bootstrapStyle }: AlertProps): ReactElement | null {
    return message ? <div className={classNames(`alert alert-${bootstrapStyle}`, className)}>{message}</div> : null;
}
