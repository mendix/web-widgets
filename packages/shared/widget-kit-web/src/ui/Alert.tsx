import classNames from "classnames";
import { createElement, ReactNode, Children } from "react";

export interface AlertProps {
    children?: ReactNode;
    className?: string;
    bootstrapStyle: "default" | "primary" | "success" | "info" | "warning" | "danger";
}

export function Alert({ className, bootstrapStyle, children }: AlertProps): JSX.Element | null {
    return Children.count(children) > 0 ? (
        <div className={classNames(`alert alert-${bootstrapStyle}`, className)}>{children}</div>
    ) : null;
}
