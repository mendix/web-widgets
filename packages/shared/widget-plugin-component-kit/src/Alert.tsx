import { Children, createElement, ReactNode } from "react";
import classNames from "classnames";

export interface AlertProps {
    children?: ReactNode;
    className?: string;
    bootstrapStyle: "default" | "primary" | "success" | "info" | "warning" | "danger";
    role?: string;
}

export interface ValidationAlertProps {
    children?: ReactNode;
    className?: string;
}

// cloning from https://gitlab.rnd.mendix.com/appdev/appdev/-/blob/master/client/src/widgets/web/helpers/Alert.tsx
export const ValidationAlert = ({ className, children }: ValidationAlertProps): JSX.Element | null => (
    <Alert className={classNames("mx-validation-message", className)} bootstrapStyle="danger" role="alert">
        {children}
    </Alert>
);

export const Alert = ({ className, bootstrapStyle, children, role }: AlertProps): JSX.Element | null =>
    Children.count(children) > 0 ? (
        <div className={classNames(`alert alert-${bootstrapStyle}`, className)} role={role}>
            {children}
        </div>
    ) : null;

Alert.displayName = "Alert";
