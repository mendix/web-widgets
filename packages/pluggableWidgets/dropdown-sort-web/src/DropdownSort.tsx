import { Context, createElement, ReactElement, useCallback, useContext, useMemo, useRef } from "react";
import { ListAttributeValue } from "mendix";
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { SortContextValue, SortDirection, SortInstruction, getSortDispatcher } from "@mendix/widget-plugin-sorting";
import { DropdownSortContainerProps } from "../typings/DropdownSortProps";
import { SortComponent, SortOption } from "./components/SortComponent";

export function DropdownSort(props: DropdownSortContainerProps): ReactElement {
    const id = useRef(`DropdownSort${generateUUID()}`);

    const SortContext = (getSortDispatcher() ?? {}) as Context<SortContextValue>;
    const alertMessage = (
        <Alert bootstrapStyle="danger">
            The Drop-down sort widget must be placed inside the header of the Gallery widget.
        </Alert>
    );
    const { sortDispatcher, attributes, initialSort } = useContext(SortContext) ?? {};

    let defaultValue;
    let defaultDirection;

    const options = useMemo(
        () => [
            {
                caption: props.emptyOptionCaption?.value ?? "",
                value: ""
            },
            ...(attributes?.map(attribute => ({ caption: attribute.caption, value: attribute.attribute.id })) ?? [])
        ],
        [attributes, props.emptyOptionCaption?.value]
    );

    const updateSort = useCallback(
        (value: SortOption, direction: SortDirection): void => {
            const filteredOption = attributes?.find(attr => attr.attribute.id === value.value);
            sortDispatcher({
                getSortCondition: () => getSortCondition(filteredOption?.attribute, direction, value)
            });
        },
        [attributes, sortDispatcher]
    );

    if (!sortDispatcher || !attributes) {
        return alertMessage;
    }

    if (initialSort && initialSort.length > 0) {
        // Keeping the code in order to implement a future multi-sorting
        const [sort] = initialSort;
        const [id, direction] = sort;
        if (id) {
            defaultValue = options.find(option => option.value === String(id));
        }
        if (direction) {
            defaultDirection = direction;
        }
    }

    return (
        <SortComponent
            key={defaultValue?.value ?? "sort-component"}
            className={props.class}
            defaultDirection={defaultDirection}
            defaultOption={defaultValue}
            emptyOptionCaption={props.emptyOptionCaption?.value}
            id={id.current}
            options={options}
            screenReaderButtonCaption={props.screenReaderButtonCaption?.value}
            screenReaderInputCaption={props.screenReaderInputCaption?.value}
            styles={props.style}
            tabIndex={props.tabIndex}
            updateSort={updateSort}
        />
    );
}

function getSortCondition(
    listAttribute: ListAttributeValue | undefined,
    direction: SortDirection,
    value: SortOption
): SortInstruction | undefined {
    if (!listAttribute || !listAttribute.sortable || !value) {
        return undefined;
    }

    const { id } = listAttribute;

    return [id, direction];
}
