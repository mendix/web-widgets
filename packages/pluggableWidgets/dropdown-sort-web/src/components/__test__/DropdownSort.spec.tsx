import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { SortInstruction, SortContextValue } from "@mendix/widget-plugin-sorting";
import { ListAttributeValueBuilder } from "@mendix/widget-plugin-test-utils";
import { createContext, createElement } from "react";
import { DropdownSort } from "../../DropdownSort";
import { render } from "@testing-library/react";
import { mount } from "enzyme";
import { SortComponent } from "../SortComponent";

const commonProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
    sortOrder: "asc" as const
};

describe("Dropdown Sort", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with correct context", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext({
                    sortDispatcher: jest.fn(),
                    attributes: [
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withSortable(true)
                                .build(),
                            caption: "Option 1"
                        },
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withSortable(true)
                                .build(),
                            caption: "Option 2"
                        }
                    ]
                } as SortContextValue);
            });

            it("loads correct values from attributes", () => {
                const filter = mount(<DropdownSort {...commonProps} />);

                expect(filter.find(SortComponent).prop("options")).toStrictEqual([
                    {
                        caption: "",
                        value: ""
                    },
                    {
                        caption: "Option 1",
                        value: "attribute1"
                    },
                    {
                        caption: "Option 2",
                        value: "attribute2"
                    }
                ]);
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DropdownSort {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            });
        });

        describe("with view state", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext({
                    sortDispatcher: jest.fn(),
                    attributes: [
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withSortable(true)
                                .build(),
                            caption: "Option 1"
                        },
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withSortable(true)
                                .build(),
                            caption: "Option 2"
                        }
                    ],
                    initialSort: [["attribute2", "asc"] as SortInstruction]
                } as SortContextValue);
            });

            it("loads correct default option", () => {
                const filter = mount(<DropdownSort {...commonProps} />);

                expect(filter.find(SortComponent).prop("defaultOption")).toStrictEqual({
                    caption: "Option 2",
                    value: "attribute2"
                });
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            });

            it("renders error message", () => {
                const filter = mount(<DropdownSort {...commonProps} />);

                expect(filter.find(Alert).text()).toBe(
                    "The Drop-down sort widget must be placed inside the header of the Gallery widget."
                );
            });
        });
    });

    describe("with multiple instances", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext({
                sortDispatcher: jest.fn(),
                attributes: [
                    {
                        attribute: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("String")
                            .withSortable(true)
                            .build(),
                        caption: "Option 1"
                    },
                    {
                        attribute: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("Decimal")
                            .withSortable(true)
                            .build(),
                        caption: "Option 2"
                    }
                ]
            } as SortContextValue);
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(<DropdownSort {...commonProps} />);
            const { asFragment: fragment2 } = render(<DropdownSort {...commonProps} />);

            console.warn(fragment1().querySelector("input")?.getAttribute("aria-controls"));

            expect(fragment1().querySelector("input")?.getAttribute("aria-controls")).not.toBe(
                fragment2().querySelector("input")?.getAttribute("aria-controls")
            );
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });
});
