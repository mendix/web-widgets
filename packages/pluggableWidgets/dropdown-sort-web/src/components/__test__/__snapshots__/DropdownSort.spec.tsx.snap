// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Sort with single instance with correct context renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="dropdown-triggerer-wrapper"
    >
      <input
        aria-controls="DropdownSort1-dropdown-list"
        aria-expanded="false"
        aria-haspopup="true"
        class="form-control dropdown-triggerer"
        value="Option 1"
      />
      <button
        class="btn btn-default btn-sort icon-asc"
      />
    </div>
  </div>
</DocumentFragment>
`;
